---
title: "Class 15 Lab Report"
author: 'Bangyan Hu (PID: A15540189)'
date: "11/16/2021"
output: pdf_document
---

\#1. Bioconductor and DESeq2 setup

```{r}
#install.packages("BiocManager")
#BiocManager::install()
#BiocManager::install("DESeq2")
library(BiocManager)
library(DESeq2)
```

\#2. Import countData and colData

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

```{r}
head(counts)
```

```{r}
head(metadata)
```

***Q1.*** How many genes are in this dataset?

```{r}
nrow(counts)
```

There are 38694 genes in this dataset.

***Q2.*** How many 'control' cell lines do we have?

```{r}
sum(metadata$dex == "control")
```

We have 4 control cell lines.

\#3. Toy differential gene expression

First I need to extract all the "control" columns. Then, I will take the rowwise mean to get the average count values for all genes in these four experiments.

```{r}
control.ins <- metadata$dex == "control"
control.counts <- counts[ ,control.ins]
control.mean <- rowSums( control.counts )/4 
head(control.mean)
```

Alternatively

```{r}
library(dplyr)
control <- metadata %>% filter(dex=="control")
control.counts <- counts %>% select(control$id) 
control.mean <- rowSums(control.counts)/4
head(control.mean)
```

***Q3.*** How would you make the above code in either approach more robust?

To make the above code in either approach more robust, firstly, I need to extract all the "control" columns. Then, I will take the row-wise mean to get the average count values for all genes in these four experiment.

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[, control.inds]
head(control.counts)
```
```{r}
control.mean <- rowMeans(control.counts)
head(control.mean)
```
Thus, we can just use the equation given above so we can get the average of the rows even when we add more control or treated groups, making the code in either approach more robust.

***Q4.*** Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

Now do the same for the drug treated experiments

```{r}
treated.ins <- metadata$dex == "treated"
treated.counts <- counts[ ,treated.ins]
treated.mean <- rowSums( treated.counts )/4 
head(treated.mean)
```

```{r}
meancounts <- data.frame(control.mean, treated.mean)
colSums(meancounts)
```

***Q5 (a).*** Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

```{r}
plot(meancounts[,1],meancounts[,2], xlab="Control", ylab="Treated")
```

***Q5 (b).***You could also use the ggplot2 package to make this figure producing the plot below. What geom\_?() function would you use for this plot?

geom_point()

```{r}
library(ggplot2)
ggplot(meancounts, aes(meancounts[,1],meancounts[,2], xlab="Control", ylab="Treated")) + geom_point(alpha = 0.4)
```


Wait a sec. There are 60,000-some rows in this data, but I'm only seeing a few dozen dots at most outside of the big clump around the origin.

This plot indicates that we need a log transformation to see details of our data.

***Q6.*** Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

log

```{r}
plot(meancounts, log="xy")
```

Use log2 for easier interpretation.

```{r}
log2(10/10)
```

```{r}
log2(40/10)
```

```{r}
log2(5/10)
```

0 values for no change and plus values for increases and minus values for decreases. This nice property leads us to work with **log2(fold-change)** all the time in the genomics and proteomics filed.

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

***Q7.*** What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The arr.ind=TRUE argument is able to clause which() to return both the row and column indices, such as positions, where there are TRUE values. So, it will tell us which genes (rows) and samples (columns) have zero counts in this case.
As we need to ignore any genes that have zero counts in any sample, we just focus on the row answer. Calling unique() function will ensure that we do not count any row twice if it has zero entries in both samples.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

***Q8.*** Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?

```{r}
sum(up.ind)
```

There are 250 up regulated genes we have at the greater than 2 fc level.

***Q9.*** Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?

```{r}
sum(down.ind)
```

There are 367 down regulated genes we have at the greater than 2 fc level.

***Q10.*** Do you trust these results? Why or why not?

For the section above, all of the analysis was based on fold change. Nonetheless, fold change can be large, for example: two-fold up- or down-regulation, without being statistically significant as based on p-values for instance. 
As we have not done anything yet to determine whether the differences we are seeing are significant, these results in their current form are likely to be very misleading. Thus, we need to do this properly with the help of the DESeq2 package.

\#4. DESeq2 analysis

```{r}
library(DESeq2)
citation("DESeq2")
```

Importing data

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

DESeq analysis

```{r}
dds <- DESeq(dds)
```

Getting results

```{r}
res <- results(dds)
res
```

```{r}
summary(res)
```

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

\#5. Adding annotation data

```{r}
#BiocManager::install("AnnotationDbi")
#BiocManager::install("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(res)
```

***Q11.*** Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res\$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

head(res)
```

Arrange and view the results by the adjusted p-value

```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])
```

```{r}
write.csv(res[ord,], "deseq_results.csv")
```

\#6. Data Visualization

Volcano plots

```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```

To make this more useful we can add some guidelines (with the abline() function) and color (with a custom color vector) highlighting genes that have padj\<0.05 and the absolute log2FoldChange\>2.

```{r}
plot(res$log2FoldChange,  -log(res$padj), 
 ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```

To color the points we will setup a custom color vector indicating transcripts with large fold change and significant differences between conditions:

```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

EnhancedVolcano for further customization.

```{r}
#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
```

```{r}
x <- as.data.frame(res)

EnhancedVolcano(x,
                lab = x$symbol,
                x = 'log2FoldChange',
                y = 'pvalue')
```

\#7. Pathway analysis

Pathway analysis with R and Bioconductor

```{r}
#BiocManager::install( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

![](hsa05310.pathview.png)

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa05310", kegg.native=FALSE)
```

***Q12.*** Can you do the same procedure as above to plot the pathview figures for the top 2 down-regulated pathways?

```{r}
## Focus on top 2 down-regulated pathways here
keggrespathwaysd <- rownames(keggres$less)[1:2]

# Extract the 8 character long IDs part of each string
keggresidsd = substr(keggrespathwaysd, start=1, stop=8)
keggresidsd
```

Do the same procedure as above to plot the pathview figures for the top 2 down-regulated pathways:

hsa04940 Type I diabetes mellitus
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04940")
```
![](hsa04940.pathview.png)


hsa05332 Graft-versus-host disease
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05332")
```
![](hsa05332.pathview.png)



OPTIONAL: Plotting counts for genes of interest

```{r}
i <- grep("CRISPLD2", res$symbol)
res[i,]
```

```{r}
rownames(res[i,])
```

```{r}
plotCounts(dds, gene="ENSG00000103196", intgroup="dex")
```

```{r}
# Return the data
d <- plotCounts(dds, gene="ENSG00000103196", intgroup="dex", returnData=TRUE)
head(d)
```

```{r}
boxplot(count ~ dex , data=d)
```

```{r}
library(ggplot2)
ggplot(d, aes(dex, count, fill=dex)) + 
  geom_boxplot() + 
  scale_y_log10() + 
  ggtitle("CRISPLD2")
```
